Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp

```diff
@@ -2422,13 +2422,14 @@ void FunctionDefinitionWriter::visit(AST::ForStatement& statement)
         m_stringBuilder.append(' ');
         visit(*update);
     }
-    m_stringBuilder.append(") "_s);
+    m_stringBuilder.append(") { volatile bool __wgslEnsureForwardProgress; (void)__wgslEnsureForwardProgress; "_s);
     visit(statement.body());
+    m_stringBuilder.append('}');
 }
 
 void FunctionDefinitionWriter::visit(AST::LoopStatement& statement)
 {
-    m_stringBuilder.append("while (true) {\n"_s);
+    m_stringBuilder.append("while (true) { volatile bool __wgslEnsureForwardProgress; (void)__wgslEnsureForwardProgress; \n"_s);
```

Comment:
The read of uninitialized variable is UB, so you're adding more UB here.
I'm not sure it makes sense to instantiate the variable inside the loop. When you add initializer, it will add a volatile write in the loop bodies. 

It's better to add the variable before the first loop, and make all nested loops use the same variable.